File: HomeWork/HomeWork12/HomeWork12.py
import heapq

def minCostToSupplyWater(n, wells, pipes):

    weighted_graph = {}
    for pipe in pipes:
        if pipe[0] not in weighted_graph:
            weighted_graph[pipe[0]] = []
        if pipe[1] not in weighted_graph:
            weighted_graph[pipe[1]] = []
        weighted_graph[pipe[0]].append([pipe[1], pipe[2]])
        weighted_graph[pipe[1]].append([pipe[0], pipe[2]])

    
    weighted_graph[0] = []
    for h in range(1, n + 1):
        if h not in weighted_graph:
            weighted_graph[h] = []
        weighted_graph[h].append([0, wells[h - 1]])
        weighted_graph[0].append([h, wells[h - 1]])

    # Prim's algorithm
    min_heap = [(0, 0)]  # (w_i, house_i)
    included = [False] * (n + 1)
    count_visited = 0
    min_cost = 0

    while count_visited <= n and min_heap:
        w, h = heapq.heappop(min_heap)
        if not included[h]:
            included[h] = True
            count_visited += 1
            min_cost += w
        else:
            continue

        for neighbor in weighted_graph[h]:
            if not included[neighbor[0]]:
                heapq.heappush(min_heap, (neighbor[1], neighbor[0]))

    return min_cost

# Example usage:
n = 3
wells = [1, 2, 2]
pipes = [[1, 2, 1], [2, 3, 1]]
print(minCostToSupplyWater(n, wells, pipes))  # Output should be 3

File: HomeWork/HomeWork12/HomeWork12.java

package HomeWork.HomeWork12;

import java.util.*;

public class HomeWork12 {
	public static int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        // house_i:[[house_j, w_ij], [house_k, w_ik],...[house_virtual, w_ii]]
        Map<Integer, List<int[]>> weightedGraph = new HashMap<>();
        // real house are numbered 1 -> n, plus '0' as the virtual house that connects 
        // to every house with weight being the cost of building a well inside of one house
        for (int[] pipe : pipes) {
            weightedGraph.computeIfAbsent(pipe[0], k -> new ArrayList<>()).add(new int[]{pipe[1], pipe[2]});
            weightedGraph.computeIfAbsent(pipe[1], k -> new ArrayList<>()).add(new int[]{pipe[0], pipe[2]});
        }

        weightedGraph.put(0, new ArrayList<>());
        for (int h = 1; h <= n; h++) {
            weightedGraph.computeIfAbsent(h, k -> new ArrayList<>()).add(new int[]{0, wells[h - 1]});
            weightedGraph.get(0).add(new int[]{h, wells[h - 1]});
        }

        // minHeap of [house_i, w_i]
        Queue<int[]> minHeap = new PriorityQueue<>((e1, e2) -> e1[1] - e2[1]);
        boolean[] included = new boolean[n + 1];
        int countVisited = 0;
        int minCost = 0;
        // pick a random starting house: 0
        minHeap.add(new int[2]);
        while (countVisited <= n && !minHeap.isEmpty()) {
            int[] edge = minHeap.remove();
            int h = edge[0], w = edge[1];
            if (!included[h]) {
                included[h] = true;
                countVisited++;
                minCost += w;
            } else {
                continue;
            }

            for (int[] neighbor : weightedGraph.get(h)) {
                if (!included[neighbor[0]]) {
                    minHeap.add(neighbor);
                }
            }
        }

        return minCost;
    }
    public static void main(String[] args) {
        
    }
}
